libgnucash/app-utils/gnc-autoclear.cpp

```diff
@@ -0,0 +1,219 @@
+#include <cstring>
+
+#define log_module "gnc.autoclear"
+
+#define MAX_SECONDS 1
+
+struct RuntimeMonitor
+{
+    uint64_t m_counter;
+    double m_seconds;
+    std::chrono::steady_clock::time_point m_start;
+    RuntimeMonitor (double seconds)
+        : m_counter(0)
+        , m_seconds(seconds), m_start(std::chrono::steady_clock::now()) {};
+    bool should_abort ()
+    {
+        if (++m_counter % 100000 != 0) return false;
+        const auto elapsed = std::chrono::duration_cast<std::chrono::duration<double>>(std::chrono::steady_clock::now() - m_start).count();
+        return elapsed > m_seconds;
+    }
+};
+
+struct SplitInfo
+{
+    int64_t m_amount;
+    Split* m_split;
+};
+
+using SplitInfoVec = std::vector<SplitInfo>;
+using SplitVec = std::vector<Split*>;
+
+static const char*
+path_to_str(const SplitInfoVec& path)
+{
+    if (path.empty()) return "<empty>";
+
+    static char buff[1000];
+    char* p = buff;
+    char* end = buff + sizeof(buff);
+    for (size_t i = 0; i < path.size(); ++i)
+    {
+        if (i > 0)
+        {
+            if (p + 1 >= end) return "<overflow>";
+            *p++ = '|';
+        }
+
+        auto [ptr, ec] = std::to_chars (p, end, path[i].m_amount);
+        if (ec != std::errc() || ptr >= end) return "<overflow>";
+        p = ptr;
+    }
+
+    p = std::min (end - 1, p);
+    *p = '\0';
+    return buff;
+}
+
+static void
+subset_sum (SplitInfoVec::const_iterator iter,
+            SplitInfoVec::const_iterator end,
+            SplitInfoVec& path, SplitVec& solution,
+            int64_t target, RuntimeMonitor& monitor)
+{
+    ENTER ("this=%" PRId64" path=%s target=%" PRId64,
+           iter == end ? 0 : iter->m_amount, path_to_str (path), target);
+
+    if (monitor.should_abort())
+    {
+        LEAVE ("exceeds time limit");
+        throw std::runtime_error ("exceeds time limit");
+    }
+
+    if (target == 0)
+    {
+        if (!solution.empty())
+        {
+            LEAVE ("abort: found another solution %s", path_to_str (path));
+            throw std::runtime_error ("The selected amount cannot be cleared.");
+        }
+
+        std::transform (path.begin(), path.end(), std::back_inserter(solution),
+                        [](SplitInfo& i){ return i.m_split; });
+
+        LEAVE ("solution found %s", path_to_str (path));
+        return;
+    }
+
+    if (iter == end)
+    {
+        LEAVE ("end of splits");
+        return;
+    }
+
+    int64_t current_num = iter->m_amount;
+    auto next = std::next(iter);
+    while (next != end && next->m_amount == current_num) ++next;
+    auto dupe_count = std::distance(iter, next);
+
+    // 1st path: use current_num
+    path.insert(path.end(), iter, next);
+    subset_sum (next, end, path, solution, target - current_num * dupe_count,
+                monitor);
+    path.erase(path.end() - dupe_count, path.end());
+
+    // 2nd path: skip current_num
+    subset_sum (next, end, path, solution, target, monitor);
+    LEAVE ("");
```

Comment:
Fwiw With this exit macro, the preceding recursive call is no longer tail call optimisable...